### DISCUSSION TITLE
Review assert() usage

### DISCUSSION THREAD

**Discussion by jonasraoni** [pkp/pkp-lib#8749] [Proposals]
**Describe the bug**
Given that an `assert()` may stop the script execution or not depending on the `php.ini` setup (e.g. `assert.exception`), it shouldn't be used for control-flow nor as a way to generate useful warnings.

```php
PHP Fatal error:  Uncaught AssertionError: assert(!empty($op)) in /var/www/html/lib/pkp/classes/core/PKPComponentRouter.php:193
Stack trace:
#0 /var/www/html/lib/pkp/classes/core/PKPComponentRouter.php(193): assert(false, 'assert(!empty($...')
#1 /var/www/html/lib/pkp/classes/core/PKPComponentRouter.php(279): PKP\core\PKPComponentRouter->getRpcServiceEndpoint(Object(APP\core\Request))
#2 /var/www/html/lib/pkp/classes/core/Dispatcher.php(163): PKP\core\PKPComponentRouter->route(Object(APP\core\Request))
#3 /var/www/html/lib/pkp/classes/core/PKPApplication.php(372): PKP\core\Dispatcher->dispatch(Object(APP\core\Request))
#4 /var/www/html/index.php(21): PKP\core\PKPApplication->execute()
#5 {main}
  thrown in /var/www/html/lib/pkp/classes/core/PKPComponentRouter.php on line 193
```

**Comment by asmecher** [comment]
Agreed, we're swapping out assertions for exceptions as a general practice. But is this issue filed for a general case, or for the specific assertion in the quoted log snippet?

**Reply by asmecher** [reply]
Any objections to swapping out assertions for Exceptions as a general practice/habit?

**Reply by jonasraoni** [reply]
@kaitlinnewson just created an issue (https://github.com/pkp/pkp-lib/issues/10285), as there are no drawbacks in this discussion, I think it can be closed.
The approach will be basically to replace by an Exception/type hint where appropriate or to remove it if not useful.

**Comment by jonasraoni** [comment]
As the `assert.*` settings fall into the `PHP_INI_ALL` group, I just wanted to check if we can define some rules to have a more uniform behavior, independent of the user configuration.

**Comment by NateWr** [comment]
I'll close this as wontfix. If you want to make a general recommendation, try the discussions. If you have a specific part of the code that is causing a bug or needs to be chnaged, you can file an issue with that.
